src/session.ts

```diff
@@ -0,0 +1,278 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import utils = require('./utils')
+import kernel = require('./kernel')
+
+import Signal = phosphor.core.Signal;
+import emit = phosphor.core.emit;
+import IAjaxSuccess = utils.IAjaxSuccess;
+import IAjaxError = utils.IAjaxError;
+
+
+/**
+ * Notebook Identification specification.
+ */
+export
+interface INotebookId {
+  path: string;
+};
+
+
+/**
+ * Session Identification specification.
+ */
+export
+interface ISessionId {
+  id: string;
+  notebook: INotebookId;
+  kernel: kernel.IKernelId;
+};
+
+
+/**
+ * Session initialization options.
+ */
+export
+interface ISessionOptions {
+  notebookPath: string;
+  kernelName: string;
+  baseUrl: string;
+  wsUrl: string;
+};
+
+
+/**
+ * Session object for accessing the session REST api. The session
+ * should be used to start kernels and then shut them down -- for
+ * all other operations, the kernel object should be used.
+ **/
+export 
+class Session {
+
+  static statusChanged = new Signal<Session, string>();
+
+  /**
+   * GET /api/sessions
+   *
+   * Get a list of the current sessions.
+   */
+  static list(sessionServiceUrl: string): Promise<ISessionId[]> {
+    return utils.ajaxRequest(sessionServiceUrl, {
+      method: "GET",
+      dataType: "json"
+    }).then((success: IAjaxSuccess): ISessionId[] => {
+      if (success.xhr.status === 200) {
+        if (!Array.isArray(success.data)) {
+          throw Error('Invalid Session list');
+        }
+        for (var i = 0; i < success.data.length; i++) {
+          validateSessionId(success.data[i]);
+        }
+        return success.data;
+      }
+      throw Error('Invalid Status: ' + success.xhr.status);
+    });
+  }
+
+  /**
+   * Construct a new session.
+   */
+  constructor(options: ISessionOptions) {
+    this._notebookModel = {
+      path: options.notebookPath
+    };
+    this._kernelModel = {
+      id: null,
+      name: options.kernelName
+    };
+
+    this._baseUrl = options.baseUrl;
+    this._wsUrl = options.wsUrl;
+    this._sessionServiceUrl = utils.urlJoinEncode(this._baseUrl, 
+                                                  'api/sessions');
+  }
+
+  /**
+   * Get the session kernel object.
+  */
+  get kernel() : kernel.Kernel {
+    return this._kernel;
+  }
+
+  /**
+   * POST /api/sessions
+   *
+   * Start a new session. This function can only executed once.
+   */
+  start(): Promise<void> {
+    return utils.ajaxRequest(this._sessionServiceUrl, {
+      method: "POST",
+      dataType: "json",
+      data: JSON.stringify(this._model),
+      contentType: 'application/json'
+    }).then((success: IAjaxSuccess) => {
+      if (success.xhr.status !== 201) {
+        throw Error('Invalid response');
+      }
+      validateSessionId(success.data);
+      this._updateModel(success.data);
+      if (this._kernel) {
+        this._kernel.name = this._kernelModel.name;
```

Comment:
My memory is fuzzy on this one, but I think it has to do with opening a notebook that is already running, in which case the server has the authoritative value for the kernelspec, since it's already running. It may not be needed anymore, and it's possible that we can come up with a cleaner separation of the two ways this stuff gets set up:

1. notebook wasn't open, new session is created, new kernel is started
2. notebook was open, session returns existing kernel

being careful to distinguish the existence server-side session/kernel objects from the instantiation of the js-side API objects for them.